packages/gatsby-plugin-graphql-config/src/gatsby-node.ts

```diff
@@ -0,0 +1,111 @@
+import * as fs from "fs-extra"
+import { resolve, join } from "path"
+import { printSchema, GraphQLSchema } from "graphql"
+import type { GatsbyReduxStore } from "gatsby/src/redux"
+import type { IStateProgram } from "gatsby/src/internal"
+
+async function cacheSchema(
+  cacheDirectory: string,
+  schema: GraphQLSchema
+): Promise<void> {
+  try {
+    const schemaSDLString = printSchema(schema, { commentDescriptions: true })
+
+    await fs.writeFile(join(cacheDirectory, `schema.graphql`), schemaSDLString)
+
+    console.log(`[gatsby-plugin-graphql-config] wrote SDL file to .cache`)
+  } catch (err) {
+    console.error(
+      `[gatsby-plugin-graphql-config] failed writing schema file to .cache`
+    )
+    console.error(err)
+  }
+}
+
+async function cacheGraphQLConfig(program: IStateProgram): Promise<void> {
+  try {
+    const base = program.directory
+    const configJSONString = JSON.stringify(
+      {
+        schema: resolve(base, `.cache/schema.graphql`),
+        documents: [
+          resolve(base, `src/**/**.{ts,js,tsx,jsx,esm}`),
+          resolve(base, `.cache/fragments.graphql`),
+        ],
+        extensions: {
+          endpoints: {
+            default: {
+              url: `${program.https ? `https://` : `http://`}${program.host}:${
+                program.port
+              }/___graphql`,
+            },
+          },
+        },
+      },
+      null,
+      2
+    )
+
+    fs.writeFileSync(
+      resolve(base, `.cache`, `graphql.config.json`),
+      configJSONString
+    )
+    console.log(`[gatsby-plugin-graphql-config] wrote config file to .cache`)
+  } catch (err) {
+    console.error(
+      `[gatsby-plugin-graphql-config] failed to write config file to .cache`
+    )
+    console.error(err)
+  }
+}
+
+const createFragmentCacheHandler = (
+  cacheDirectory: string,
+  store: GatsbyReduxStore
+) => async (): Promise<void> => {
+  try {
+    const currentDefinitions = store.getState().definitions
+
+    const fragmentString = Array.from(currentDefinitions.entries())
+      .filter(([_, def]) => def.isFragment)
+      .map(([_, def]) => `# ${def.filePath}\n${def.printedAst}`)
+      .join(`\n`)
+
+    await fs.writeFile(
+      join(cacheDirectory, `fragments.graphql`),
+      fragmentString
+    )
+
+    console.log(`[gatsby-plugin-graphql-config] wrote fragments file to .cache`)
+  } catch (err) {
+    console.error(
+      `[gatsby-plugin-graphql-config] failed writing fragments file to .cache`
+    )
+    console.error(err)
+  }
+}
+
+export async function onPostBootstrap({
```

Comment:
I think query extraction occurs during bootstrap, so we won't see the initial `SET_DEFINITIONS` event if we hook into `onPostBootstrap`. Have you tried it? Are fragments written to disk after the initial `gatsby develop` run?

If not, then I think we should start listening to emiter in `onPreBootstrap`.

Reply:
yes! the gifs were showing this in action. the `definitions` are not present `onPostBootstrap` anyways when using `store.getState()`, so the emitter is what causes the fragments to appear, sometime after bootstrap.